### ISSUE TITLE
Fonts is broken in firefox's pdf reader

### ISSUE THREAD

**Issue by TatriX** [fschutt/printpdf#20]
I just did
```
cargo run --example fonts
```
and then
```
firefox test_fonts.pdf
```
and the file is broken.

The same happens to us in production with external partner's pdf reader software.

**Comment by fschutt** [comment]
I could reproduce it - the file doesn't work in Firefox, but it does work in:

Evince:

![image](https://user-images.githubusercontent.com/12084016/43585047-678c6530-9664-11e8-830a-aa86713dd44c.png)

Adobe Reader:

![image](https://user-images.githubusercontent.com/12084016/43584791-b2ae90a2-9663-11e8-9e84-33babf438056.png)

Foxit Reader:

![image](https://user-images.githubusercontent.com/12084016/43585004-4886fb6e-9664-11e8-9cfd-a9cf8e32e189.png)

xpdf:

![image](https://user-images.githubusercontent.com/12084016/43585634-e631052a-9665-11e8-92f1-691a0adf5564.png)

This leads me to believe that there is nothing wrong with the file itself but rather with Firefox' PDF reader,  especially since I ran the fonts example through Adobe Distiller until all checks passed. I can run it again, but as far as I know, it validates as a PDF.

What third-party software are you using?

**Comment by fschutt** [comment]
What is very noticable is that Firefox does display the page boundaries. If the file was truly "broken", Firefox wouldn't be able to do that (it would just show a blank screen). So I think that https://github.com/mozilla/pdf.js can't decode Unicode character widths correctly and just displays a generic "file is broken" error.

**Comment by fschutt** [comment]
On the other hand, xpdf gives me some warnings:

```
Syntax Warning: Mismatch between font type and embedded font file
```

and https://www.pdf-online.com/osa/validate.aspx also complains about the font format:

```
Validating file "test_fonts.pdf" for conformance level pdf1.3
    The embedded font program 'F0' cannot be read.
    The document contains fonts without embedded font programs or encoding information (CMAPs).

Done.
```

**Comment by TatriX** [comment]
We have problems with non-ascii characters. The problem with third-party software is that we have no idea what they are using to read our pdfs.

**Comment by fschutt** [comment]
@TatriX Do you have screenshots of the error that they get or other error information, tickets, email conversation, anything? Could you ask them what software they are using (even if it's proprietary third-party software, at least the name could help me out). 

"Problems with non-ascii characters": The fonts.rs example only uses ASCII characters, so I assume that you didn't send them the fonts.rs example, but they have problems with some different PDF. Could you ask them to send the exact PDF that your clients have problems with? And if they can view ASCII characters, I assume they get a different error than just "PDF corrupt", since they can load the PDF (?)

If there is confidential information, send an email to felix.schuett@maps4print.com, I'll keep any information private of course.

I'll happily fix the error, it's just that "problems with non-ascii characters" is a very non-descriptive error and I have nothing to work with. It could be that the problem is the reason why Firefox doesn't display the page contents, but it could also be a different error.

**Comment by TatriX** [comment]
At the moment we switched to wkhtmltopdf. I'll ask if I'm allowed to send you the docs and will do that if that's the case. Thank you for your fast response and the project itself!

**Comment by fschutt** [comment]
I've forwarded this to https://github.com/mozilla/pdf.js/issues/9949 (at least for the Firefox issue). While I tried debugging this, I don't understand the pdf.js architecture well enough and Adobe Preflight doesn't give me any errors.

I'll notify you if the bug gets fixed (if it is a problem with this library). But still, thanks for the report.

**Comment by jrmuizel** [comment]
I had a look at the CMap for test_fonts.pdf and it looks wrong.

As excerpted below. 


```
1 beginbfchar
<0000> <0000>
<0001> <0000>
<0002> <0002>
<0003> <000d>
<0004> <0020>
<0005> <0021>
<0006> <0022>
<0007> <0023>
...
```

The parameter to beginbfchar is '1' when it should be the number of characters that follow.

From the CMap spec:


> The beginbfchar and endbfchar operators map int number of individual input codes (srcCode) to a corresponding number of individual character codes (dstCode) or character names (dstCharname), where int can be ≤ 100.

**Comment by fschutt** [comment]
Damn, thanks @jrmuizel. I thought it was the ID of the block, because the example PDFs I worked from used "1" and "2" as numbers. That's quite a large mistake on my part, I'll fix that soon

**Comment by fschutt** [comment]
@jrmuizel While it's now fixed (in 4a1a701d46e9aaf56de44707c98f3867caa185cb), it sadly doesn't fix the Firefox rendering issue.

**Comment by fschutt** [comment]
@TatriX Firefox now displays the PDF without problems:

![image](https://user-images.githubusercontent.com/12084016/43660685-00f8dca2-9760-11e8-862d-77b776aa7bbe.png)

The reason was that the type of the embedded font was marked incorrectly (`CIDFontType0` instead of the correct `CIDFontType2`). Other PDF viewers are a bit more lenient there. This was fixed in 209bb910a19dc37fd7eac2d6c60c14f6387b84fb. Could you try again if this fixes your issue? Otherwise, I regard this issue as closed for good.

**Comment by TatriX** [comment]
Works fine now, thanks!
